Assets/Src/Scripts/AgileMvvm/ViewModel.cs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace AgileMvvm {
+  // The ViewModel class that maintains models and raises model update events to the bound Views. It
+  // also provides the utility methods to manage the binding relationships.
+  //
+  // In AgileMvvm, the relationships among View, ViewModel and Model are:
+  //
+  // * The Model does not know about the View or the ViewModel.
+  // * The View does not know about the Model, but knows the ViewModel.
+  // * The View may own the ViewModel, but it is not necessary to do that. The View can communicate
+  //   with the ViewModel via lightweight bindings even when it doesn't own the ViewModel.
+  // * The binding relationships between the View and the ViewModel can be set up either by the View
+  //   itself, or by the Application class, the GameManager class, etc.
+  // * Different Views or multiple View instances can be bound to the same ViewModel.
+  // * The ViewModel does not know about the View, but knows and maintains the Model.
+  //
+  // Typical MVVM patterns support the following binding modes between the View and the ViewModel:
+  //
+  // * One-way binding: ViewModel's property to View's property or View's event handler.
+  // * One-way-to-source binding: View's property to ViewModel's property. (Rarely used)
+  // * Two-way binding: Two-way update between View's property and ViewModel's property.
+  // * Event binding: View's event to ViewModel's event handler.
+  //
+  // As an lightweight framework, AgileMvvm explicitly supports the one-way binding mode via the
+  // ViewModel class. To establish a one-way binding between the View and the ViewModel:
+  //
+  // * Defines your own ViewModel class and implement bindable properties like below:
+  //
+  //    class Person : ViewModel {
+  //      private string _name;
+  //      [BindableProperty]
+  //      public string Name {
+  //        get => _name;
+  //        set => MvvmSetter(ref _s, value);
+  //      }
+  //    }
+  //
+  // * Defines your View class:
+  //
+  //    class Dialog {
+  //      public string TextField { get; set; }
+  //    }
+  //
+  // * Binds the bindable property of the ViewModel to the target property of the View class:
+  //
+  //    var dialog = new Dialog();
+  //    var person = new Person();
+  //    person.Bind("Name", dialog, "TextField");
```

Comment:
There isn't compile time check for the string name of the properties. Is it possible to implement like "Bind(person.Name, dialog.TextField)" in C#? I have no idea about it now.

Reply:
It seems no perfect way to do this.

(1) A solution (also used by many implementations including `Microsoft.Toolkit.Mvvm`) is C#'s `nameof` expression. With `nameof` expressions, the interface of the Bind method will still being look like `Bind(string propertyName...)`, accepting string parameters. It's the caller's decision whether to use `nameof` or not.

(2) It's not a solution to simply accept `PropertyInfo` as the parameter type. This requires the caller to pass in something like `typeof(MyView).GetProperty("Foo")`, which is still depending on a compile-time string literal.

(3) Microsoft WPF and `Microsoft.Toolkit.Mvvm` also use XML to bind Views and ViewModels. E.g., `SelectedItem="{x:Bind ViewModel.SelectedSubreddit, Mode=TwoWay}"`. Technically, this is still a compile-time string literal solution.

I updated the sample code, the test code and the SeedCalc code to the `nameof` solution for now. It looks a bit better than before.

Let me know if you think of any better ideas.

Reply:
I don't have better ideas, we can keep current design now.